### PR TITLE
sock: Introduce xmpp_sock_t abstraction

### PR DESCRIPTION
**Pull request by pasis**
**This patch needs review and testing!**

libstrophe uses non-blocking sockets and the connect() syscall may return before a TCP connection is established. This doesn't allow to catch all possible errors synchronously and some of the errors are handled in the event handler.

In a scenario with multiple SRV records and/or multiple IP addresses resolution, we need to repeat connection attempt on a failure.

xmpp_sock_t resolves the above problem. It keeps resolved records and addresses to repeat connect attempt asynchronously.

### CODE REVIEW

**Comment by sjaeckel** [comment]
CC @nosnilmot

File: src/conn.c
Comment (nosnilmot) [inline src/conn.c:1858]: Moving the invocation of `sockopt_cb()` out of `sock_connect()` means user can no longer use this callback to set any socket options before `connect()`, was it necessary to do this?

File: src/conn.c
Comment (nosnilmot) [inline src/conn.c:962]: On Windows sock_t is unsigned:
```
../src/conn.c: In function 'conn_disconnect':
../src/conn.c:962:20: warning: comparison of unsigned expression in '>= 0' is always true [-Wtype-limits]
  962 |     if (conn->sock >= 0)
      |                    ^~
```

suggest instead `if (conn->sock != INVALID_SOCKET)` and for non-Windows add `#define INVALID_SOCKET -1` inside `#ifndef _WIN32` in `sock.h`. (`INVALID_SOCKET` is already defined by Winsock)

File: src/event.c
Comment (nosnilmot) [inline src/event.c:67]: On Windows sock_t is unsigned:
```
../src/event.c: In function '_connect_next':
../src/event.c:67:20: warning: comparison of unsigned expression in '< 0' is always false [-Wtype-limits]
   67 |     if (conn->sock < 0)
```

File: src/conn.c
Comment (nosnilmot) [inline src/conn.c:1852]: On Windows sock_t is unsigned:
```
../src/conn.c: In function '_conn_connect':
../src/conn.c:1852:24: warning: comparison of integer expressions of different signedness: 'sock_t' {aka 'unsigned int'} and 'int' [-Wsign-compare]
 1852 |         if (conn->sock == -1)
```

File: src/sock.c
Comment (nosnilmot) [inline src/sock.c:225]: On Windows sock_t is unsigned:
```
../src/sock.c: In function 'sock_connect':
../src/sock.c:225:19: warning: comparison of unsigned expression in '< 0' is always false [-Wtype-limits]
  225 |     } while (sock < 0);
      |                   ^
```

**Review by nosnilmot — COMMENTED** [review]
This is looking good and I've verified it handles failures well in the tests I've done. Just a few minor points.

File: src/conn.c
Comment (pasis) [reply src/conn.c:1852]: This is something new to me since I don't have experience with windows. I guess I saw other places in the code which need to be fixed as well. Will fix it.

File: src/conn.c
Comment (pasis) [reply src/conn.c:1858]: Good point, I completely missed this. The idea was to remove xmpp_conn_t knowledge from the sock component. I will return old implementation back for now.

File: src/event.c
Comment (pasis) [reply src/event.c:67]: Will fix it.

File: src/sock.c
Comment (pasis) [reply src/sock.c:225]: Will fix it

File: src/conn.c
Comment (pasis) [reply src/conn.c:962]: Will fix it,

File: src/conn.c
Comment (pasis) [reply src/conn.c]: done

File: src/event.c
Comment (pasis) [reply src/event.c]: done

File: src/conn.c
Comment (pasis) [reply src/conn.c]: done

**Comment by pasis** [comment]
@nosnilmot I've resolved unsigned socket issue in all places I found. And returned user setsockopt functionality back before connect() syscall. Could you please check current patch again?

**Comment by fantomfp** [comment]
[libstrophe-profanity-dev-profrc.txt](https://github.com/strophe/libstrophe/files/11781347/libstrophe-profanity-dev-profrc.txt)
[libstrophe-profanity-dev-debug.log](https://github.com/strophe/libstrophe/files/11781348/libstrophe-profanity-dev-debug.log)
Some infos about the test environment:

sources of libstrophe: git repo (xmpp-sock branch)

libstrophe has been built in a Fedora 38 x86_64 chroot

libstrophe has been installed on the system inside a chroot (Fedora 38 x86_64)

The Fedora RPM of libstrophe was reused (dirty RPM / dirty packaging) to manage the installation in the system

profanity was installed via RPM (profanity-0.13.1-3.fc38.x86_64)

a standard user has been created and configured inside the chroot

profanity was launched as non-root

debug log is sent in attachment as proof

WORKS FOR ME.

Please merge as soon as possible :)



Some infos about the infra:

1 jabber server used as main server. 1 server as fallback.

DNS zone is configured to always redirect clients to the main server. If it doesn't answer, clients will try to connect to the secondary server.

```
$ dig SRV +short _xmpp-client._tcp.casperlefantom.net.
10 0 5222 jabber2.casperlefantom.net.
5 0 5222 jabber.casperlefantom.net.
```

For the tests, I turned off the main server (ejabberd) during few minutes. All clients got disconnected. Then, they reconnected on the fallback server.

For the /reconnect settings used during the tests, I'm adding my profrc in attachement.

XMPP account is a test account.

**Review by nosnilmot — APPROVED** [review]
all my previous comments have been addressed 😄

**Review by sjaeckel — APPROVED** [review]
Thanks! :+1: very nice

Waiting on the re-review of @nosnilmot before merging.

**Pull request merged**
